sdk/pinocchio/src/sysvars/instructions.rs

```diff
@@ -0,0 +1,287 @@
+/// specified index.
+/// 
+/// # Errors
+/// 
+/// Returns [`ProgramError::UnsupportedSysvar`] if the given account's ID is not equal to [`ID`].
+/// Returns [`ProgramError::InvalidArgument`] if the index is out of bounds.
+#[inline(always)]
+pub fn load_instruction_at_checked(
+    index: usize,
+    instruction_sysvar_account_info: &AccountInfo,
+) -> Result<IntrospectedInstruction, ProgramError> {
+    if instruction_sysvar_account_info.key() != &INSTRUCTIONS_ID {
+        return Err(ProgramError::UnsupportedSysvar);
+    }
+
+    let instruction_sysvar_data = instruction_sysvar_account_info.try_borrow_data()?;
+    unsafe {
+        load_instruction_at(index, &instruction_sysvar_data)
+            .map(|instr| instr.clone())
+            .map_err(|err| match err {
+                SanitizeError::IndexOutOfBounds => ProgramError::InvalidArgument,
+                _ => ProgramError::InvalidInstructionData,
+            })
+    }
+}
+
+/// Returns the `Instruction` relative to the current `Instruction` in the
+/// currently executing `Transaction`.
+///
+/// # Errors
+///
+/// Returns [`ProgramError::UnsupportedSysvar`] if the given account's ID is not equal to [`ID`].
+/// Returns [`ProgramError::InvalidArgument`] if the index is out of bounds.
+pub fn get_instruction_relative(
+    index_relative_to_current: i64,
+    instruction_sysvar_account_info: &AccountInfo,
+) -> Result<IntrospectedInstruction, ProgramError> {
+    if instruction_sysvar_account_info.key() != &INSTRUCTIONS_ID {
+        return Err(ProgramError::UnsupportedSysvar);
+    }
+
+    let current_index = load_current_index_checked(&instruction_sysvar_account_info)? as i64;
+    let index = current_index.saturating_add(index_relative_to_current);
+    if index < 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let instruction_sysvar_data = instruction_sysvar_account_info.try_borrow_data()?;
+    unsafe {
+        load_instruction_at(index as usize, &instruction_sysvar_data)
+            .map(|instr| instr.clone())
+            .map_err(|err| match err {
+                SanitizeError::IndexOutOfBounds => ProgramError::InvalidArgument,
+                _ => ProgramError::InvalidInstructionData,
+            })
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, PartialEq, Eq)]
+pub struct IntrospectedInstruction {
+    pub raw: *const u8,
+}
+
+impl IntrospectedInstruction {
+    /// Get the account meta at the specified index.
+    /// 
+    /// # Safety
+    /// 
+    /// This function is unsafe because it does not verify if the index is out of bounds.
+    /// 
+    /// It is typically used internally within the `get_account_meta_at` function, which
+    /// performs the necessary index verification. However, to optimize performance for users 
+    /// who are sure that the index is in bounds, we have exposed it as an unsafe function.
+    pub unsafe fn get_account_meta_at_unchecked(&self, index: usize) -> IntrospectedAccountMeta {
```

Comment:
Making the change to `IntrospectedAccountMeta`, we can return a reference to the account directly:
```
 pub unsafe fn get_account_meta_at_unchecked(&self, index: usize) -> &IntrospectedAccountMeta {
    let offset = core::mem::size_of::<u16>() + (index * IntrospectedAccountMeta::LEN);
    &*(self.raw.add(offset) as *const IntrospectedAccountMeta)
}
```